src/daft-csv/src/read.rs

```diff
@@ -54,57 +77,159 @@ async fn read_csv_single(
+    max_chunks_in_flight: Option<usize>,
+    estimated_mean_row_size: Option<usize>,
 ) -> DaftResult<Table> {
+    let (schema, estimated_mean_row_size) = match schema {
+        Some(schema) => (schema.to_arrow()?, estimated_mean_row_size),
+        None => {
+            let (schema, estimated_mean_row_size) = read_csv_schema_single(
+                uri,
+                has_header,
+                Some(delimiter),
+                // Read at most 1 MiB when doing schema inference.
+                Some(1 << 20),
+                io_client.clone(),
+                io_stats.clone(),
+            )
+            .await?;
+            (schema.to_arrow()?, Some(estimated_mean_row_size))
+        }
+    };
+    let compression_codec = CompressionCodec::from_uri(uri);
     match io_client
         .single_url_get(uri.to_string(), None, io_stats)
         .await?
     {
         GetResult::File(file) => {
-            read_csv_single_from_reader(
-                File::open(file.path).await?.compat(),
+            read_csv_from_compressed_reader(
+                BufReader::new(File::open(file.path).await?),
+                compression_codec,
                 column_names,
                 include_columns,
                 num_rows,
                 has_header,
                 delimiter,
+                schema,
+                // Default buffer size of 512 KiB.
+                buffer_size.unwrap_or(512 * 1024),
+                // Default chunk size of 64 KiB.
+                chunk_size.unwrap_or(64 * 1024),
+                // Default max chunks in flight would result in 2 * 8 * 1024 * chunk_size bytes, or 1 GiB with the default chunk size.
+                max_chunks_in_flight.unwrap_or(8 * 1024),
+                // If no estimated row size information from schema inference, we assume 200 bytes per row.
+                estimated_mean_row_size.unwrap_or(200),
             )
             .await
         }
-        result @ GetResult::Stream(..) => {
-            // TODO(Clark): Enable streaming remote reads by wrapping the BoxStream in a buffered stream that's
-            // (1) sync and (2) seekable.
-            read_csv_single_from_reader(
-                Cursor::new(result.bytes().await?),
+        GetResult::Stream(stream, _, _) => {
+            read_csv_from_compressed_reader(
+                StreamReader::new(stream),
+                compression_codec,
                 column_names,
                 include_columns,
                 num_rows,
                 has_header,
                 delimiter,
+                schema,
+                // Default buffer size of 512 KiB.
+                buffer_size.unwrap_or(512 * 1024),
+                // Default chunk size of 64 KiB.
+                chunk_size.unwrap_or(64 * 1024),
+                // Default max chunks in flight would result in 2 * 8 * 1024 * chunk_size bytes, or 1 GiB with the default chunk size.
+                max_chunks_in_flight.unwrap_or(8 * 1024),
```

Comment:
Why is this number so high given that the parallelism is limited by the number rayon tasks that can run in parallel which would be limited by the number of cores on a machine. Then the max number in flight could just be the `num_cpus` on the current machine?

Reply:
Opted for 2x # of cores so we can (1) pipeline I/O-bound reading with compute-bound parsing, and (2) try to ensure that each worker thread will have a downloaded chunk immediately available for parsing even if parsing is completing in ~uniform waves.